src/app/Dashboard/Charts/MBeanMetricsChartController.tsx

```diff
@@ -0,0 +1,220 @@
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import { ApiService } from '@app/Shared/Services/Api.service';
+import { SettingsService } from '@app/Shared/Services/Settings.service';
+import { Target, TargetService } from '@app/Shared/Services/Target.service';
+import {
+  BehaviorSubject,
+  catchError,
+  concatMap,
+  distinctUntilChanged,
+  finalize,
+  first,
+  map,
+  merge,
+  Observable,
+  of,
+  pairwise,
+  ReplaySubject,
+  Subject,
+  Subscription,
+  tap,
+  throttleTime,
+} from 'rxjs';
+
+export class MBeanMetricsChartController {
+  private readonly _metrics = new Map<string, string[]>();
+  private readonly _state$ = new Subject<MBeanMetrics>();
+  private readonly _loading$ = new BehaviorSubject<boolean>(true);
+  private readonly _refCount$ = new BehaviorSubject<number>(0);
+  private readonly _updates$ = new ReplaySubject<void>(1);
+  private readonly _lazy: Subscription;
+  private _attach: Subscription | undefined;
+
+  constructor(
+    private readonly _api: ApiService,
+    private readonly _target: TargetService,
+    private readonly _settings: SettingsService
+  ) {
+    this._lazy = this._refCount$
+      .pipe(
+        map((v) => v > 0),
+        pairwise()
+      )
+      .subscribe(([prev, curr]) => {
+        if (!prev && curr) {
+          // first subscriber joined
+          this._start();
+        }
+        if (prev && !curr) {
+          // last subscriber left
+          this._stop();
+        }
+      });
+  }
+
+  attach(category: string, fields: string[]): Observable<MBeanMetrics> {
+    this._refCount$.next(this._refCount$.value + 1);
+    if (!this._metrics.has(category)) {
+      this._metrics.set(category, []);
+    }
+    this._metrics.get(category)?.push(...fields);
+    return this._state$.asObservable().pipe(
+      distinctUntilChanged(),
+      finalize(() => {
+        this._refCount$.next(this._refCount$.value - 1);
+        const original = this._metrics.get(category) || [];
+        const updated = [...original];
+        fields.forEach((field) => {
+          const idx = updated.findIndex((e) => e === field);
+          if (idx >= 0) {
+            updated.splice(idx, 1);
+          }
+        });
+        this._metrics.get(category)?.push(...updated);
+      })
+    );
+  }
+
+  requestRefresh(): void {
+    this._loading$.next(true);
```

Comment:
Seems like in the `_attach` obs chain we have a tap to set this already. We can remove it here?